File: MarkdownEditor/MarkdownEditor/Sources/Editor/EditorDialogPresenter.m
//
//  EditorDialogPresenter.m
//  MarkdownEditor
//

#import "EditorDialogPresenter.h"
#import <UniformTypeIdentifiers/UniformTypeIdentifiers.h>

@implementation EditorDialogPresenter

- (void)showAlertWithTitle:(NSString *)title
                   message:(NSString *)message
                 forWindow:(NSWindow *)window {
    if (!NSThread.isMainThread) {
        dispatch_sync(dispatch_get_main_queue(), ^{
            [self showAlertWithTitle:title message:message forWindow:window];
        });
        return;
    }

    NSAlert *alert = [[NSAlert alloc] init];
    alert.informativeText = title;
    alert.messageText = message;
    alert.alertStyle = NSAlertStyleWarning;
    [alert addButtonWithTitle:@"OK"];
    [alert beginSheetModalForWindow:window completionHandler:nil];
}

- (void)confirmDiscardingChangesForWindow:(NSWindow *)window
                        completionHandler:(void (^)(NSModalResponse returnCode))handler {
    NSAlert *alert = [[NSAlert alloc] init];
    alert.informativeText = @"Your changes will be lost if you don't save them.";
    alert.messageText = @"Do you want to save the changes you made to New file?";
    alert.alertStyle = NSAlertStyleWarning;
    [alert addButtonWithTitle:@"Save"];
    [alert addButtonWithTitle:@"Cancel"];
    [alert addButtonWithTitle:@"Don't Save"];
    [alert beginSheetModalForWindow:window completionHandler:handler];
}

- (void)showOpenFilePanelForWindow:(NSWindow *)window
                       initialPath:(NSString *)initialPath
                 completionHandler:(void (^)(NSURL * _Nullable selectedURL))handler {
    NSOpenPanel *panel = [NSOpenPanel openPanel];
    panel.directoryURL = [NSURL fileURLWithPath:initialPath];
    panel.canChooseDirectories = NO;
    panel.canChooseFiles = YES;
    panel.canSelectHiddenExtension = YES;
    [panel beginSheetModalForWindow:window completionHandler:^(NSModalResponse result) {
        handler(result == NSModalResponseOK ? panel.URL : nil);
    }];
}

- (void)showSaveFilePanelForWindow:(NSWindow *)window
                       initialPath:(NSString *)initialPath
                 completionHandler:(void (^)(NSURL * _Nullable selectedURL))handler {
    NSSavePanel *panel = [NSSavePanel savePanel];
    panel.directoryURL = [NSURL fileURLWithPath:initialPath];
    NSMutableArray<UTType *> *allowedContentTypes = [NSMutableArray array];
    UTType *mdType = [UTType typeWithFilenameExtension:@"md"];
    UTType *markdownType = [UTType typeWithFilenameExtension:@"markdown"];
    if (mdType) {
        [allowedContentTypes addObject:mdType];
    }
    if (markdownType) {
        [allowedContentTypes addObject:markdownType];
    }
    panel.allowedContentTypes = allowedContentTypes;
    panel.canSelectHiddenExtension = YES;
    [panel beginSheetModalForWindow:window completionHandler:^(NSModalResponse result) {
        handler(result == NSModalResponseOK ? panel.URL : nil);
    }];
}

@end

File: MarkdownEditor/MarkdownEditor/Sources/Editor/EditorDialogPresenter.swift
import Cocoa
import UniformTypeIdentifiers

@objc(EditorDialogPresenter)
class EditorDialogPresenter: NSObject {
    @objc(showAlertWithTitle:message:forWindow:)
    func showAlert(title: String, message: String, for window: NSWindow?) {
        let alert = warningAlert(title: title, message: message)
        if let window {
            alert.beginSheetModal(for: window)
        }
    }

    private func warningAlert(title: String, message: String) -> NSAlert {
        let alert = NSAlert()
        alert.informativeText = title
        alert.messageText = message
        alert.alertStyle = .warning
        alert.addButton(withTitle: "OK")
        return alert
    }

    @objc(confirmDiscardingChangesForWindow:completionHandler:)
    func confirmDiscardingChanges(for window: NSWindow?, completionHandler handler: @escaping (NSApplication.ModalResponse) -> Void) {
        let alert = NSAlert()
        alert.informativeText = "Your changes will be lost if you don't save them."
        alert.messageText = "Do you want to save the changes you made to New file?"
        alert.alertStyle = .warning
        alert.addButton(withTitle: "Save")
        alert.addButton(withTitle: "Cancel")
        alert.addButton(withTitle: "Don't Save")
        if let window {
            alert.beginSheetModal(for: window, completionHandler: handler)
        } else {
            handler(.cancel)
        }
    }

    @objc(showOpenFilePanelForWindow:initialPath:completionHandler:)
    func showOpenFilePanel(for window: NSWindow?, initialPath: String, completionHandler handler: @escaping (URL?) -> Void) {
        let panel = NSOpenPanel()
        panel.directoryURL = URL(fileURLWithPath: initialPath)
        panel.canChooseDirectories = false
        panel.canChooseFiles = true
        panel.canSelectHiddenExtension = true
        if let window {
            panel.beginSheetModal(for: window) { result in
                handler(result == .OK ? panel.url : nil)
            }
        } else {
            handler(nil)
        }
    }

    @objc(showSaveFilePanelForWindow:initialPath:completionHandler:)
    func showSaveFilePanel(for window: NSWindow?, initialPath: String, completionHandler handler: @escaping (URL?) -> Void) {
        let panel = NSSavePanel()
        panel.directoryURL = URL(fileURLWithPath: initialPath)
        panel.allowedContentTypes = ["md", "markdown"].compactMap { UTType(filenameExtension: $0) }
        panel.canSelectHiddenExtension = true
        if let window {
            panel.beginSheetModal(for: window) { result in
                handler(result == .OK ? panel.url : nil)
            }
        } else {
            handler(nil)
        }
    }
}
